topology/pod-cidr/src/main/java/org/apache/hadoop/net/PodCIDRToNodeMapping.java

```diff
@@ -0,0 +1,307 @@
+import com.google.common.collect.Sets;
+import com.google.common.net.InetAddresses;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import io.fabric8.kubernetes.api.model.Node;
+import io.fabric8.kubernetes.api.model.NodeList;
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.ConfigBuilder;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.utils.HttpClientUtils;
+import okhttp3.Dispatcher;
+import okhttp3.OkHttpClient;
+import org.apache.commons.cli.BasicParser;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.commons.net.util.SubnetUtils;
+import org.apache.commons.net.util.SubnetUtils.SubnetInfo;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.log4j.BasicConfigurator;
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+
+/**
+ * A namenode topology plugin mapping pods to cluster nodes for a K8s configured with pod CIDR.
+ *
+ * For each k8s pod, determines a network path with three components. The full path would look like:
+ *
+ *     RACK-NAME '/' NODE-NAME '/' POD-HOST
+ *
+ * , where NODE-NAME is the cluster node that the pod is running on.
+ *
+ * To comply with this, datanodes will be also put the node name into the same hierarchy.
+ *
+ *     RACK-NAME '/' NODE-NAME '/' NODE-NAME
+ *
+ * This way, the namenode will see the datanode and pods in the same node are closer than otherwise.
+ *
+ * The resolve method below only returns the first parts for input entries.
+ *
+ * Note this three level hierarchy requires NetworkTopologyWithNodeGroup to be used in namenode.
+ * To use this plugin, add the followings to the hdfs-site.xml:
+ * <pre>
+ * {@code
+ * <property>
+ *   <name>net.topology.node.switch.mapping.impl</name>
+ *   <value>org.apache.hadoop.net.PodCIDRToNodeMapping</value>
+ * </property>
+ * <property>
+ *   <name>net.topology.impl</name>
+ *   <value>org.apache.hadoop.net.NetworkTopologyWithNodeGroup</value>
+ * </property>
+ * <property>
+ *   <name>net.topology.nodegroup.aware</name>
+ *   <value>true</value>
+ * </property>
+ * <property>
+ *   <name>dfs.block.replicator.classname</name>
+ *   <value>org.apache.hadoop.hdfs.server.blockmanagement.BlockPlacementPolicyWithNodeGroup</value>
+ * </property>
+ * }
+ * </pre>
+ */
+@SuppressWarnings("unused")
+public class PodCIDRToNodeMapping extends AbstractDNSToSwitchMapping {
+
+  private static final String DEFAULT_NETWORK_LOCATION = NetworkTopology.DEFAULT_RACK +
+      NetworkTopologyWithNodeGroup.DEFAULT_NODEGROUP;
+
+  private static Log log = LogFactory.getLog(PodCIDRToNodeMapping.class);
```

Comment:
I think this is technically mutable and using the lower case is a slightly better practice.